src/DataFrames.jl

```diff
@@ -101,6 +101,17 @@ else
     export only
 end
 
+"""
+    DataFrames.MULTITHREADING
+
+This global flag controls whether DataFrames functions may spawn tasks that can run
+on multiple threads in parallel when deemed appropriate.
+This is set to `true` by default.
+Call `DataFrames.MULTITHREADING[] = false` to disable multithreading,
+for example because distribution of work across threads is managed separately.
+"""
+const MULTITHREADING = Threads.Atomic{Bool}(true)
```

Comment:
maybe set it to `Threads.nthreads() > 1`? I.e. when there is only a single thread we disable multi-threading? Intuitively it should be a bit faster (but I have not looked at the code below yet, so maybe it does not matter)

Reply:
Actually I tried to do that so that we don't need to also check `Threads.nthreads() > 1`, but I realized that it would be suboptimal to spawn tasks if for some reason users set `MULTITHREADING` to `true` when `Threads.nthreads() == 1`. And this could happen if somebody does `DataFrames.MULTITHREADING[] = false; ... something thread-unsafe... DataFrames.MULTITHREADING[] = true`.

Reply:
Agreed, I am just asking about what the default value for this variable should be when DataFrames.jl is loaded.

Reply:
OK. Then I tend to think always using `true` by default is simpler. The performance gain you refer to would be due to not having to check `Threads.nthreads() > 1`? That's probably negligible, right?

Reply:
in line 264 (comment below) - you do not check for this. Maybe add a check and then we are OK as you propose.

Reply:
(or maybe what we have is equivalent it seems)

Reply:
I strongly recommend not exposing `MULTITHREADING::Threads.Atomic{Bool}` as an API. I also suggest clearly separating interactive ("REPL-only") UI that is called by humans and APIs that can be called from programs. My suggestion would be something like the following:

```julia
# Internals:
const INTERACTIVE_SINGLETHREADING = Threads.Atomic{Bool}(false)
const SINGLETHREADING_REGION_DEPTH = Threads.Atomic{Int}(0)

# Maybe API?
ismultithreading() = SINGLETHREADING_REGION_DEPTH[] == 0

# API
function singlethreading(f)
    Threads.atomic_add!(INTERACTIVE_SINGLETHREADING, 1)
    try
        return f()
    finally
        Threads.atomic_sub!(INTERACTIVE_SINGLETHREADING, 1)
    end
end

# Interactive UI
function enable_multithreading()
    if !INTERACTIVE_SINGLETHREADING[]
        INTERACTIVE_SINGLETHREADING[] = true
        Threads.atomic_add!(INTERACTIVE_SINGLETHREADING, 1)
    end
end

function disable_multithreading()
    if INTERACTIVE_SINGLETHREADING[]
        INTERACTIVE_SINGLETHREADING[] = false
        Threads.atomic_sub!(INTERACTIVE_SINGLETHREADING, 1)
    end
end
```

This ensures that `f()` in `DataFrames.singlethreading(f)` observes only `ismultithreading() == false` even if `DataFrames.singlethreading` is used concurrently.

Reply:
@tkf Thank you for a comment. There seem to be some typos in your code (mixed up `INTERACTIVE_SINGLETHREADING` and `SINGLETHREADING_REGION_DEPTH`). However, even if these would be fixed could you please comment what you want to achieve through this API:

My understanding is that you want the following:
1. Functions should check if they can use multi-threading using `ismultithreading()` function
2. if someone calls `diasble_multithreading()` thin it guarantees that multithreading will not be used by any function until `enable_multithreading()` is called.
3. even if global multithreading is enabled one can disable multi-threading locally when calling some function by writing `singlethreading(f)(args...; kwargs...)`. I understand that by this you want to ensure that changing the global flag concurrently does not affect the `singlethreading(f)` function call once it has been started.

My problem with understanding the API proposal is that if you call `singlethreading(f)(args...; kwargs...)` you disable multi-threading also in other functions that would be called when still `singlethreading(f)` is being executed. Then such functions would observe `ismultithreding()` to be `false` when they start but then it could turn to `true` when these other functions are still executed.

Therefore the question is, if it would not be better to:
* have a global variable only (accessible via getter/setter - here I agree) for the time being;
* wait till Julia-wide solution for passing contexts is accepted and then make use of it (instead of implementing a partial solution that is DataFrames.jl specific).

(so in general my thinking is that I would ideally not define advanced mechanism in DataFrames.jl but rather delegate this task to a general Julia solution for this).

Reply:
Sorry, there were mistakes in the variable names. I think I fixed them.

> if you call `singlethreading(f)(args...; kwargs...)`

Not sure if this is a typo, but it'd be `singlethreading(() -> f(args...; kwargs...))` instead. Or, I think people would use do block:

```julia
singlethreading() do
    # not multi-threaded
    some_dataframes_jl_api(df)
    another_dataframes_jl_api(df)
end
```

> Then such functions would observe `ismultithreding()` to be `false` when they start but then it could turn to `true` when these other functions are still executed.

This is not possible unless you leak tasks with something like

```julia
@sync singlethreading() do
    @async some_dataframes_jl_api(df)  # may see `ismultithreding() == true`
end
```

But this is rather awful.

> * wait till Julia-wide solution for passing contexts is accepted and then make use of it

I _think_ it is highly likely you can just replace the definition of `singlethreading` to use the Julia API instead. But this depends on how other Julia devs feel about the context variables API.

All that said, of course, if you prefer a simple implementation, I think that's a reasonable engineering decision. But, if you go this direction, I think you'd want to clarify that this does not work:

> And this could happen if somebody does `DataFrames.MULTITHREADING[] = false; ... something thread-unsafe... DataFrames.MULTITHREADING[] = true`.

if multiple pieces of code can potentially do this. My suggestion is a simple way to support this without ContextVariablesX.jl.

Reply:
FWIW I think there's a very high chance that `enable_multithreading`/`disable_multithreading` cannot be implemented using context variables. The API proposed in ContextVariablesX.jl does not support this to maximize the chance that context variables can be added to Julia. That's why I mentioned the separation of the UI and the API. If you want to provide a single interface and want to avoid breaking changes, `singlethreading(f)` is the safest option.

Anyway, my initial comment "strongly recommend" was clearly too strong and has a lot of assumptions that may or may not be correct. A global flag `MULTITHREADING` may be a decent solution.

Reply:
@nalimilan - no matter what we decide I think we should have a getter/setter interface and not ask users to manually modify the variable. In this way later we will be able to change the mechanics how the getter/setter works without being breaking.

Reply:
@krynju @jpsamaroo Could you confirm whether a syntax like `singlethreading() do` would be OK for DTable? I.e. instead of setting a global flag once, you'd wrap blocks that call table operations (including notably DataFrames) in such a call to disable multithreading. Ideally we will later use this approach in other packages too (for now it would have no effect on them, though it wouldn't throw an error either).

Reply:
If you want to avoid low-level stuff in DataFrames.jl, I'm also happy to create a small package (say) ThreadingHints.jl that does what I suggested above if DataFrames.jl simply wants to use `ThreadingHints.ismultithreaded()` (or maybe `prefermultithreaded()` or something). End users can then use `ThreadingHints.singlethreading(f, yes::Bool = true)`.

Reply:
@jpsamaroo + @tkf

What @nalimilan and I think is the following:
* we want to support a mechanism that Dagger.jl would use;
* if @jpsamaroo says he would not use what @tkf proposes (or something similar) anyway for now then we will define some simpler mechanism for getting/setting the flag if multi-threding should be used and wait for some generally accepted solution and use it in the future.

---

In summary: @jpsamaroo - could you please comment what kind of mechanism would you use if it were made available now?

Our (on DataFrames.jl side) only restriction is that if your answer is that you would use something experimental then we will probably not add it until it becomes non-experimental (DataFrames.jl is past 1.0 release so we need to be super careful what we add to the package as later we cannot make breaking changes).

I hope our intentions are clear. Thank you!

Reply:
The problem is that adding support for toggling arbitrary global flags in Dagger leads to unpredictable and non-composable behavior (because now other unrelated code sees the effect of Dagger changing this flag, and we also need to track the value of the flag per-worker, which adds extra overhead and complexity to Dagger).

Basically, in the absence of a composable mechanism (i.e. context variables), it's on the user to change this flag to something that is preferred. That's not a big deal, it's just something we need to document and communicate.

I'm sorry for not pointing this all out earlier.

Reply:
Let me add some context to my decision with an example:

A user has a continuous processing pipeline that uses DataFrames. They're very happy with it, and the new DataFrames multithreading feature has made it fast and efficient.

This user now wants to add an HTTP server to this process that takes some of the results of this pipeline in real-time and serves them to clients. To do this, they use Dagger to spawn off some pre-processing operations using the `DTable`, which will produce the results that will be sent to the requesting client.

Now, they tested this new piece in isolation with test data, and it worked out really well, and the `DTable` effectively turned off DataFrames multithreading to ensure that the pre-processing tasks for HTTP clients ran efficiently.

With both pieces working well in isolation, they combine the codes together and run them in the same session (because they don't need a bunch of overhead from serializing data between workers). Suddenly, their continuous processing pipeline is struggling to perform well (because it's now running single-threaded), while the HTTP clients are being (quickly and efficiently) served stale data.

After some investigation, they find out what Dagger is doing, and are unhappy that simply using Dagger for data pre-processing slows down and effectively breaks their whole application. They file an issue on Dagger, and I have to point out that because of a certain design decision, this behavior is "expected". They are unhappy and stop using Dagger, opting to instead use some other approach. Dagger loses an interested user, and the user loses the benefits of Dagger's DTable.

Reply:
@jpsamaroo - I fully understand that (and kind of expected - that is why I have asked for an explicit information from your side as we really want DataFrames.jl to play nicely with Dagger.jl in the long run). The problem is that even the `singlethreading` is not "strong enough" to give you what you want as it is using a shared counter (@tkf - please corret me if I am wrong).

Therefore what I propose to do:
* for now use DataFrames.jl specific global getter/setter of multi threading that is intended to be handled on user's side
* when we have a fully working and stable context passing solution in Dagger.jl we will add functionality that will respect this context (we will decide how to do it when the context-passing solution will be implemented in Dagger.jl)

OK?

Reply:
That all sounds good to me!

Once context variables are used in DataFrames (either via ContextVariablesX or some Base-provided API) I will be happy to have Dagger propagate these variables (although I may add a new package that users can load to turn this on automatically, because this propagation still should be opt-in to be safe; automatically propagating a pointer, for example, would be very bad :smile:).

Reply:
As I said, I designed `singlethreaded(f)` specifically to make it comptaible with context variable. The implementation can be switched to use a context variable once it's available. In fact, it can also be implemented using ContextVariablesX right now. This is exactly why I suggested separating UI and API (or ditching the enable/diable UI is also OK).